### ISSUE TITLE
FR Funsor-based Pyro backend

### ISSUE THREAD

**Issue by eb8680** [pyro-ppl/pyro#2580]
This issue is for tracking Funsor integration with Pyro in the form of a new `pyroapi` backend that shares lots of code with Pyro and lives in `pyro.contrib.funsor`, as implemented in the original PR #2307 and NumPyro PR https://github.com/pyro-ppl/numpyro/pull/572 and recently enabled by the removal of Python 3.5 support from PyTorch 1.6.

The first version of `pyro.contrib.funsor` hews very closely to Pyro's existing design and semantics, including replicating parallel enumeration and plate semantics exactly, preserving `trace` and `replay`, and making extensive use of Pyro's current test suite.  I'm hoping to progressively refine away much of this legacy design after the first version is merged.

Tentative sequence of PRs expected based on the contents of #2307 (mostly done already there):
- [x]  #2583 `make_handler` change, `funsor[torch]` optional dependency, and initial `pyro.contrib.funsor` library creation: `primitives`, `to_funsor`, `to_data`, `markov`, `pyroapi`, and `test_named.py`.
- [x] #2584 First introductory tutorial
- [x] #2588 `GlobalNamedMessenger` and a second tutorial
- [x] #2589 Exact-only `EnumMessenger`, `trace`, `replay`, and `test_valid_models.py`
- [x] #2593 Vectorized `PlateMessenger` and `test_valid_models_plate.py`
- [x] #2593 Support for subsampling and `pyro.subsample` in new `PlateMessenger`, and tests
- [x] #2594 Sequential `PlateMessenger` and `test_valid_models_sequential_plate.py`
- [x] #2595 `TraceTMC_ELBO`, sampling enumeration strategies, and `test_tmc.py`.
- [x] #2596 `TraceEnum_ELBO` and `test_enum.py`

Remaining tasks to wrap up the first version of `contrib.funsor` after #2596 is merged:
- [x] #2583 Support missing or globally inconsistent input specifications in `pyro.to_funsor` and `pyro.to_data`
- [x] #2583 xfailing `test_pyroapi.py`
- [x] #2610 Update domain syntax
- [x] #2611 JIT support for `TraceEnum_ELBO` and `TraceTMC_ELBO`
- [x] #2611 missing ELBOs necessary to make `test_pyroapi.py` pass (e.g. add a simple `Trace_ELBO`)
- [x] #2614 Port some existing examples e.g. HMMs to `examples/contrib/funsor/` (this should be as simple as using `pyroapi` to switch backends)
- [x] #2663 Increase minimum PyTorch version to 1.6 and drop Python 3.5 support in Pyro before release

Some changes to consider for the next version, after everything above has been addressed:
- Rewrite inference algorithms like `TraceEnum_ELBO` entirely in terms of Funsor expressions
- Remove all internal dependence on legacy Pyro metadata: `BroadcastMessenger`, `CondIndepStackFrame`s, `Trace`, `Queue`, etc.
- Experiment with reimplementing MCMC, especially HMC components, in a more idiomatic manner
- Refactor `DimStack` to support a wider class of enumeration scenarios, such as within a DAG defined at runtime
- Add enumeration tests that are less tightly coupled to Pyro's existing enumeration machinery and work more directly with `to_funsor`/`to_data`
- Make sequential enumeration more tightly integrated with `contrib.funsor` and less visible
- Refactor enumeration strategies into `Reparametrizer`s as used in `poutine.reparam`, or something similar
- Consider changing the semantics of sequential plates to allow local enumeration
- Profile inference algorithms relative to their Pyro counterparts, with and without the PyTorch JIT
- Add better code documentation once legacy stuff has been removed
- Add type annotations and look into adding static type checking to CI
- Add a tutorial on implementing inference algorithms, perhaps something as simple as importance sampling
- Add error checking and enable test_eum.py's `assert_error` and `assert_warning` tests (see #2606)

Other tasks which are less urgent:
- Consider moving some of `funsor.pyro` into `pyro.contrib.funsor`, especially the distributions
- More complete support for converting distributions to funsors, especially `TransformedDistribution`s
- Move the `_get_support_value` helper added in #2589 upstream to Funsor
- Support for sequential enumeration in fancy elbos
- Move `pyroapi.register_backend` call upstream to pyroapi?
- Investigate any suspicious slowdowns in tests or examples

**Comment by eb8680** [comment]
I'm going to close this issue as resolved, since the current contents of `contrib.funsor` and the 1.5 release are a clear milestone. I'll create new master issues for future work, including for the speculative to-do list in this issue.
